### ISSUE TITLE
OpenSC no longer follows token label standard

### ISSUE THREAD

**Issue by CendioOssman** [OpenSC/OpenSC#1612]
### Problem Description

OpenSC sets a token label that doesn't follow PKCS#11 2.11 Appendix D, yet uses the same format. This confuses software (like our ThinLinc) that assumes adherence to that standard and causes incorrect output for the user.

From the standard:

> The tokenLabel value is generated from a combination of the real value and the PIN label. The format is "\<token label\> (\<pin label\>)" (i.e. "Electronic ID (PIN1)").

OpenSC used to follow this, but it was changed in 22f9ad704a4d69a7cb08f92987413cfd633f7a81.

And example card here, which with the old OpenSC showed:

```
$ pkcs11-tool  --list-token-slots
Available slots:
Slot 0 (0x1): SCM Microsystems Inc. SCR 3311 [CCID Interface] (21120608441518)
  token label        : Telia EID IP5a (identification)
...
```

now shows:

```
$ pkcs11-tool  --list-token-slots
Available slots:
Slot 0 (0x0): SCM Microsystems Inc. SCR 3311 [CCID Interface] (21120608441518)
  token label        : identification (Telia EID IP5a)
...
```

Front-end software following the standard will now present the user with a very unhelpful prompt asking for the "Telia EID IP5a" PIN and the user won't know if the "identification" or "signing" PIN should be used.

### Proposed Resolution

Ideally, revert 22f9ad704a4d69a7cb08f92987413cfd633f7a81 and nag Mozilla that their display is cropping of useful things and they should do something different.

If you insist on the current order, then at least use a format that cannot be confused with the standard format specified in PKCS#11. E.g. using `[]` instead of `()`.

**Comment by frankmorgner** [comment]
I can't find this appendix, please give me a link (ideally for the version that OpenSC claims to be conforming to).

I'll leave the nagging to you.

**Comment by CendioOssman** [comment]
You have to look in the version 2.11 unfortunately:

https://www.cryptsoft.com/pkcs11doc/STANDARD/pkcs-11v2-11r1.pdf

It was removed in 2.20 without any comment about a replacement method for handling multiple PINs. Most libraries still follow the original principle though, even you guys until that commit. :)

**Comment by muktupavels** [comment]
How that is Mozilla bug / problem? Token label max length is 32 and pull request of your linked commit shows same problem with `pkcs11-tool`. You can not avoid truncation if `token label` + `pin label` is larger than 32, no? And looking at https://github.com/OpenSC/OpenSC/pull/732 pin label is more important.

**Comment by frankmorgner** [comment]
I wasen't aware that the fix is so simple. Truncating longer labels leaving the PIN identifier intact should be the way to go.

**Comment by mouse07410** [comment]
I think #732 got it wrong. Most tokens have only one PIN that user can enter, so differentiating them is less important than providing the correct label for the token itself. The PKCS#11 way "Token label (PIN label)" is good.

**Comment by metsma** [comment]
Most tokens? depends
Here Estonia, Finland, Latvia, Belgium? ... have least 2 PIN-s on token and it users get confused PIN type is truncated.

**Comment by mouse07410** [comment]
@metsma you're saying that with your token a user would have different PINs for, e.g., AUTH and DIGITAL SIG keys???

**Comment by metsma** [comment]
Yes

**Comment by muktupavels** [comment]
If total length is 32 or less then "Token label (PIN label)" should not be problem, if it longer then maybe truncating like this would be better? "Very long token la... (PIN label)"?

**Comment by frankmorgner** [comment]
@muktupavels yes

**Comment by muktupavels** [comment]
Please check my comment here - https://github.com/OpenSC/OpenSC/pull/1617#issuecomment-469058943. I could redo that to switch back to "Token label (PIN label)" format.

Initially I wanted to fix truncated labels for my card - `User PIN (LATVIA I)` and `Signature PIN (LATVIA I)`. Should be `User PIN (LATVIA ID)` and `Signature PIN (LATVIA ID)` with current format.

**Comment by martinpaljak** [comment]
IIRC it was changed a long time ago exactly for the same reason (PIN label being trunkated away) so this should be reverted, indeed.

**Comment by CendioOssman** [comment]
I guess this got fixed in b7ec7f95b1b38636a1b58bd276dd68632c791209 and this issue should be closed?
